### PR TITLE
Add quantile regression with pinball loss

### LINKED ISSUE #3456: FR: quantile regression
### Summary

To regress quantiles, not just a single value. 

Ref https://arxiv.org/abs/2202.05265

### Rationale

Quantiles provides bounds

### Implementation

Lift and shift from https://github.com/lightning-uq-box/lightning-uq-box

CC @nilsleh 

### Alternatives

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by quinnarnold**
### Summary

Add `PinballLoss` and `loss="pinball"` to `Regression` and `PixelwiseRegression`. Models predict one channel per requested quantile for one regression target. The loss trains all quantiles; median predictions feed existing metrics and plots. These estimates are not calibrated prediction intervals.

Quantiles are registered as a buffer so they move with the loss module. The loss rejects mismatched prediction channels or target shapes before broadcasting. Task branches explicitly check `loss == 'pinball'`; pixelwise model channels are named separately from the constructor's target count.

### Rationale

Closes #3456. A possible application is biomass mapping with lower, median, and upper conditional quantiles. This PR does not provide field-study or production performance results.

### Verification

- `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 python -m pytest -p pytest_socket -p pytest_cov.plugin -q tests/losses/test_quantile.py tests/tasks/test_regression.py --cov=torchgeo/losses --cov=torchgeo/tasks` — 77 passed, 33 slow tests deselected. Both changed production modules have 100% line coverage.
- Numerical loss/gradient tests cover scalar and spatial targets; task tests cover quantile ordering, median metrics, and fit/test/predict. The buffer/dtype assertion fails on its previous implementation. Five invalid-shape cases fail before the shape fix and pass afterward.
- CPU YAML `fast_dev_run` fit/test passed with actual ResNet18 and U-Net/ResNet18 models using `cyclone.yaml` and `copernicus_biomass_s3.yaml`, `loss=pinball`, and repository synthetic fixtures. These are smoke runs, not a trained application study.
- CPU checks confirm ordinary `Regression(num_outputs=2)` works with MAE/MSE. Current upstream pixelwise models hard-code one output channel; that pre-existing limitation remains unchanged.
- Ruff check/format and full ty pass; `git diff --check` passes.

Read the Docs currently encounters the two missing Tokenizers references tracked in #4125, outside this PR.

### Checklist

- [x] I have read the [Contributing docs](https://docs.torchgeo.org/en/stable/user/contributing.html)
- [x] I have read the [AI policy](https://github.com/torchgeo/governance/blob/main/AI-POLICY.md)
- [ ] 🟢 **No AI usage**: written by humans, for humans
- [x] 🟡 **AI-assisted**: AI helped with the coding, but I understand every line
- [ ] 🔴 **AI-generated**: AI did everything, review with caution

### CODE REVIEW

**Comment by robmarkcole** [comment]
@quinnarnold missing tests: `57.57% of diff hit (target 100.00%)`

**Comment by quinnarnold** [comment]
Added tests in e36197ff for pinball loss values and gradients, scalar/spatial targets, invalid configuration, quantile ordering, median metrics, and the existing fit/test/predict paths. All 88 focused tests pass locally, every executable line added by this PR is covered, and Ruff/ty pass. CI is running.

Copilot's non-pinball output-channel finding is a real pre-existing issue: `PixelwiseRegression(model='fcn', num_outputs=2)` emits one channel, and a training step with a two-channel target fails the metric shape check. Upstream `main` already hard-codes `classes=1` in every pixelwise model constructor. This PR preserves that behavior outside pinball mode, so I have left multi-target pixelwise regression for a separate fix rather than expanding this single-target quantile change.

**Review by robmarkcole — DISMISSED** [review]
LGTM. Potentially we could also exercise pinball through one representative YAML-backed fast_dev_run per task type, e.g. using copernicus_biomass_s3? @calebrob6 can you also review/advise?

File: torchgeo/tasks/regression.py [hunk 1]
```diff
@@ -253,48 +293,51 @@ def configure_models(self) -> None:
...
-                    classes=1,
+                    classes=num_outputs,
                 )
             case 'segformer':
                 self.model = smp.Segformer(
                     encoder_name=backbone,
                     encoder_weights='imagenet' if weights is True else None,
                     in_channels=in_channels,
-                    classes=1,
+                    classes=num_outputs,
```
Comment (calebrob6) [inline torchgeo/tasks/regression.py:333]: Did `num_outputs` != 1 ever work before? Do the other losses work with `num_outputs > 1`

File: torchgeo/tasks/regression.py [hunk 2]
```diff
@@ -253,48 +293,51 @@ def configure_models(self) -> None:
         model = self.hparams['model']
         backbone = self.hparams['backbone']
         in_channels = self.hparams['in_channels']
+        num_outputs = (
```
Comment (calebrob6) [inline torchgeo/tasks/regression.py:296]: How is this `num_outputs` different than the  one in the constructor?

File: torchgeo/tasks/regression.py [hunk 3]
```diff
@@ -218,6 +255,8 @@ def test_step(self, batch: Sample, batch_idx: int, dataloader_idx: int = 0) -> N
             y = y.unsqueeze(dim=1)
         loss = self.criterion(y_hat, y)
         self.log('test_loss', loss, batch_size=batch_size)
+        if self.median_index is not None:
```
Comment (calebrob6) [inline torchgeo/tasks/regression.py:258]: Not a huge fan of custom loss code in the trainer -- also won't this mis-align y with y_hat? Can we move this into the actual loss function computation?

File: tests/tasks/test_regression.py [hunk 4]
```diff
@@ -230,8 +234,9 @@ def create_model(*args: Any, **kwargs: Any) -> Module:
             'copernicus_biomass_s3',
         ],
     )
+    @pytest.mark.parametrize('pinball', [False, True])
```
Comment (calebrob6) [inline tests/tasks/test_regression.py:237]: This seems _fine_ for right now but something we might want to revisit if we start accumulating more specific losses

**Review by calebrob6 — CHANGES_REQUESTED** [review]

**Comment by calebrob6** [comment]
@quinnarnold are you using this in your work? Can you speak to what the pinball loss would be good for in geoml settings? Also, have you tried training a model with this?

**Comment by blaz-r** [comment]
I believe this loss was useful for some sea gauge regression or river discarge, but I can't find any papers. So it'd be nice to hear the experience here.

File: torchgeo/losses/quantile.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        Args:
+            predictions: Predictions of shape (B, Q, ...) for Q quantiles.
+            target: Regression targets of shape (B, 1, ...).
+
+        Returns:
+            Mean pinball loss.
+        """
+        quantiles = predictions.new_tensor(self.quantiles).view(
+            1, -1, *([1] * (predictions.ndim - 2))
+        )
```
Comment (blaz-r) [inline torchgeo/losses/quantile.py]: I think we could avoid data copy here on every call and make a quantiles buffer? Loss is getting called quite often, and it seems a bit unnecessary. Could this instead become a buffer, and only the dtype and view stay inside forward? I think since loss is part of module it will be moved to device only once then.

File: torchgeo/tasks/regression.py [hunk 6]
```diff
@@ -92,7 +114,11 @@ def configure_models(self) -> None:
         weights = self.weights
         self.model = timm.create_model(
             self.hparams['model'],
-            num_classes=self.hparams['num_outputs'],
+            num_classes=(
+                len(self.hparams['quantiles'])
+                if self.median_index is not None
```
Comment (blaz-r) [inline torchgeo/tasks/regression.py]: This checks if we are using pinball loss, correct? Shouldn't the condition be more clear instead of some transitive paramter that only gets set in case of pinball loss.

File: torchgeo/tasks/regression.py (hunk 1 above)
Comment (quinnarnold) [reply torchgeo/tasks/regression.py]: Checked current main and ran CPU checks: ordinary Regression passes num_outputs through to timm, and both MAE and MSE work with two targets. PixelwiseRegression already hard-codes one channel for every model, so num_outputs > 1 was not implemented there before this PR. This change retains that existing behavior and restricts pinball to one target.

File: torchgeo/tasks/regression.py (hunk 2 above)
Comment (quinnarnold) [reply torchgeo/tasks/regression.py]: The constructor argument counts regression targets; this local value counts model channels. For pinball, one target has one channel per quantile. Renamed the local variable to num_channels and clarified the constructor docstring in cb9b6750.

File: torchgeo/tasks/regression.py (hunk 3 above)
Comment (quinnarnold) [reply torchgeo/tasks/regression.py]: This slice selects a point prediction for metrics after the loss is computed; the pinball calculation itself is in PinballLoss. The loss receives (B, Q, ...) predictions and broadcasts the (B, 1, ...) target across quantiles. The subsequent median slice keeps that singleton channel, so metrics receive matching (B, 1, ...) tensors.

I kept the slice outside the loss so PinballLoss keeps the usual scalar-return contract and all quantile channels contribute gradients. The numerical task tests cover both scalar/spatial targets, a median in a non-central channel, all-quantile gradients, and matching train/validation/test metrics. All pass. The mode check is now explicit rather than inferred from median_index.

File: tests/tasks/test_regression.py (hunk 4 above)
Comment (quinnarnold) [reply tests/tasks/test_regression.py:237]: Agreed; kept the existing parametrization. The representative cyclone and copernicus_biomass_s3 YAML smoke runs also pass with real CPU models and the repository fixtures.

File: torchgeo/losses/quantile.py (hunk 5 above)
Comment (quinnarnold) [reply torchgeo/losses/quantile.py]: Done in cb9b6750: quantiles are now a registered buffer that moves with the module. Forward only casts the dtype and reshapes it for broadcasting. Added a buffer/dtype regression that fails on the previous implementation; the focused suite passes.

File: torchgeo/tasks/regression.py (hunk 6 above)
Comment (quinnarnold) [reply torchgeo/tasks/regression.py]: Done in cb9b6750. Model construction and metric selection now explicitly check self.hparams['loss'] == 'pinball', rather than inferring the loss mode from median_index.

File: torchgeo/losses/quantile.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+import torch
+from torch import Tensor, nn
+
+
+class PinballLoss(nn.Module):
+    """Pinball loss, averaged over quantiles, samples, and spatial dimensions.
+
+    See equation (14) of `Image-to-Image Regression with Distribution-Free
+    Uncertainty Quantification <https://arxiv.org/abs/2202.05265>`_.
```
[the remark this answers is not in the document]
Comment (quinnarnold) [reply torchgeo/losses/quantile.py:16]: Covered in tests/losses/test_quantile.py: known loss values, gradients, spatial broadcasting, invalid quantiles, and now buffer/dtype movement. tests/tasks/test_regression.py also checks quantile ordering and median metrics. The current run passes 72 tests, with 100% line coverage for PinballLoss and the regression task module.

**Comment by quinnarnold** [comment]
Use cases include depth or distance estimation, biomass and crop-yield prediction, and river-discharge forecasting. I've used it for distance estimations.

For training evidence, I ran CPU YAML `fast_dev_run` fit and test passes with actual ResNet18 and U-Net/ResNet18 models for `cyclone` and `copernicus_biomass_s3`, using pinball loss and the repository's synthetic fixtures. These verify the training path, not real-data accuracy or uncertainty quality. That also covers the representative YAML smoke runs suggested by @robmarkcole.

The revision in `cb9b6750` registers quantiles as a buffer and makes the pinball condition/output-channel naming explicit. All 72 focused loss/regression tests pass; both production modules have 100% line coverage.

**Comment by robmarkcole** [comment]
My own use is crop yield, but it's applicable to any regression task

File: torchgeo/losses/quantile.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+            target: Regression targets of shape (B, 1, ...).
+
+        Returns:
+            Mean pinball loss.
+        """
+        quantiles = self.quantiles.to(dtype=predictions.dtype).view(
+            1, -1, *([1] * (predictions.ndim - 2))
+        )
+        error = target - predictions
+        return torch.maximum(quantiles * error, (quantiles - 1) * error).mean()
```
[the remark this answers is not in the document]
Comment (quinnarnold) [reply torchgeo/losses/quantile.py:63]: Fixed in `9db849af`: validate the quantile channel count and exact `(B, 1, ...)` target shape before subtraction. All five invalid-shape cases failed before the fix and now pass, including a single prediction channel and `(B,)` targets with `B == Q`. The loss/task suite passes 77 tests with 100% coverage of both changed production modules.

The separate Read the Docs failure is the missing Tokenizers references addressed by #4125.
